File: ch03/commuteatlanta-bootstrap-sampling.jl
include("utils.jl")
using GLMakie,CSV,DataFrames,Random,StatsBase,PrettyTables,Pipe,TableTransforms,Bootstrap
using Makie.Colors
desc=Lock5Table(270,"CommuteAtlanta","bootstrap sampling ",[:Distance,:Time])
data=@pipe load_data(desc.name)|>select(_,desc.feature[2])
mean_and_std(data[:,1])  #(29.11, 20.71831066456885)

#1. dotplot
function plot_dotplot(gd::GroupedDataFrame;title="CommuteAtlanta Time sampling")
    xs=@pipe keys(gd).|>values(_)[1]
    ys=[nrow(g) for g in gd]
    fig=Figure(resolution=(1200,600))
    ax=Axis(fig[1,1],xlabel="Time",ylabel="Count",title=title)
    Box(fig[1,1];color =(:orange,0.1),strokewidth=0.2)

    for (idx, x) in enumerate(xs)
        for y in 1:ys[idx]
            scatter!(ax,x,y
            ;marker=:circle,markersize=8,color=(:black)
            ,strokewidth=0.1,strokecolor=:black)
        end
    end
    
   fig
end
#groupby(data,:Time)|>plot_dotplot|>fig->save("commuteatlanta-500-sampling.png",fig)


# 2 bootstrap method

"""
    samplen(data::Array,n::Int)
    从数组 data 中随机抽取一个样本, 样本容量为 n

"""
samplen(data::Array,n::Int)=@pipe data|>Vector(1:length(_))|>data[rand(_,n)]



## basic bootstrap
n_boot = 1000   #repeat sampling times
"""
    bs(data)
    bootstraping 包装方法
"""
bs(data) = bootstrap(mean,data, BasicSampling(n_boot))

gdf=@pipe  data[:,1]|>bs|>_.t1[1]|>round.(_,digits=1)|>DataFrame(x=_)|>@orderby(_,:x)|>groupby(_,:x)

@pipe plot_dotplot(gdf;title="bootstrap-sampling")#|>save("CommuteAtlanta-Time-bootstrap-sampling.png",_)

File: marimo/ch03/commuteatlanta-bootstrap-sampling.py
import marimo

__generated_with = "0.24.0"
app = marimo.App(width="medium")


@app.cell
def _():
    import marimo as mo
    import sys
    from pathlib import Path
    import numpy as np
    import polars as pl
    import matplotlib.pyplot as plt
    from scipy import stats
    from great_tables import GT
    _p = Path(mo.notebook_location()).resolve().parent
    while _p != _p.parent and not (_p / "lock5.py").exists():
        _p = _p.parent
    if str(_p) not in sys.path:
        sys.path.insert(0, str(_p))
    import lock5
    return mo, sys, Path, np, pl, plt, stats, GT, lock5


@app.cell
def _(mo):
    mo.md(r"""# CommuteAtlanta — bootstrap sampling dot plot

Port of `UnitB/ch03/commuteatlanta-bootstrap-sampling.jl`. Bootstrap the mean
`Time` (1000 samples) and dot-plot the bootstrap means (rounded to 0.1).
Bootstrap results are random, so values differ from the Julia run.""")
    return


@app.cell
def _(lock5):
    desc = lock5.Lock5Table(270, "CommuteAtlanta", "bootstrap sampling", ["Distance", "Time"])
    data = lock5.load_data(desc.name).select(desc.feature[1])
    times = data["Time"].to_numpy().astype(float)
    return data, desc, times


@app.cell
def _(lock5, mo, np, pl, times):
    mn, sd = lock5.mean_and_std(times)
    boots = lock5.boot_sampling(times, n=1000, method=np.mean, seed=1)
    rounded = np.round(boots, 1)
    res_df = pl.DataFrame({
        "item": ["mean", "std (sample)"],
        "value": [round(mn, 2), round(sd, 3)],
    })
    mo.Html(lock5.gt_table(res_df, title="CommuteAtlanta Time").as_raw_html())
    return boots, mn, res_df, rounded


@app.cell
def _(lock5, rounded):
    fig, ax = lock5.plot_dotplot(rounded, nbins=40, title="Bootstrap sampling",
                                 xlabel="Time", ylabel="Count", color="black")
    fig
    return fig, ax


if __name__ == "__main__":
    app.run()
